File: 2018/nim/day10.nim
import sequtils, strutils, sugar

type Image = object
    row: seq[int]
    col: seq[int]
    hor_vel: seq[int]
    ver_vel: seq[int]

proc initImage(): Image =
     Image(row: newSeq[int](), col: newSeq[int](),
           hor_vel: newSeq[int](), ver_vel: newSeq[int]())

proc add_point(self: var Image, point: seq[int]) =
    self.row.add(point[0])
    self.col.add(point[1])
    self.hor_vel.add(point[2])
    self.ver_vel.add(point[3])

proc move(self: var Image, moves: int = 1) =
    for idx in 0..<len(self.row):
        self.row[idx] += self.hor_vel[idx] * moves
        self.col[idx] += self.ver_vel[idx] * moves

proc sizes(self: Image): (int, int) =
    let wid = max(self.row) - min(self.row)
    let hei = max(self.col) - min(self.col)
    (wid, hei)

proc compute_area(self: Image): int =
    let (wid, hei) = self.sizes()
    wid * hei

proc draw(self: Image) =
    let (wid, hei) = self.sizes()
    let (xmin, ymin) = (min(self.row), min(self.col))
    var render = collect(newSeq):
        for i in 1..hei + 1:
            newSeqWith(wid + 1, ' ')

    for val in zip(self.row, self.col):
        render[val[1] - ymin][val[0] - xmin] = '#'
    for row in 0..hei:
        echo render[row].join()

var image = initImage()
for line in readFile("data/input-10.txt").strip().splitLines():
    let line = line.replace("position=<", "")
        .replace("> velocity=<", ",")
        .replace(">", "")
        .replace(" ", "")
        .split(",")
        .map(parseInt)
    image.add_point(line)

var (min_area, min_iter) = (image.compute_area(), 0)
for iter in 1..11000:
    image.move()
    let area = image.compute_area()
    if area < min_area:
        (min_area, min_iter) = (area, iter)
    else:
        image.move(-1)
        break

echo "Part 1:"
image.draw()

let res2 = min_iter
echo "Part 2: ", res2
assert(res2 == 10888)

File: 2018/python/day10.py
class Image:
    def __init__(self, points):
        self.row = []
        self.col = []
        self.hor_vel = []
        self.ver_vel = []
        for row, col, hor_vel, ver_vel in points:
            self.row.append(row)
            self.col.append(col)
            self.hor_vel.append(hor_vel)
            self.ver_vel.append(ver_vel)

    def move(self, moves=1):
        for idx in range(len(self.row)):
            self.row[idx] += self.hor_vel[idx] * moves
            self.col[idx] += self.ver_vel[idx] * moves

    def __sizes(self):
        xmin, ymin = min(self.row), min(self.col)
        wid, hei = max(self.row) - xmin, max(self.col) - ymin
        return (wid, hei, xmin, ymin)

    def area(self):
        wid, hei, _, _ = self.__sizes()
        return wid * hei

    def draw(self):
        wid, hei, xmin, ymin = self.__sizes()
        render = [[" " for _ in range(wid + 1)] for _ in range(hei + 1)]
        for i, j in zip(self.row, self.col):
            render[j - ymin][i - xmin] = "#"
        for row in render:
            print("".join(row))


def parse(str):
    str = (
        str.strip()
        .replace("position=<", "")
        .replace("> velocity=<", ",")
        .replace(">", "")
    )
    row, col, hor_vel, ver_vel = map(int, str.split(","))
    return (row, col, hor_vel, ver_vel)


file = open("data/input-10.txt")
image = Image([parse(line) for line in file.readlines()])

min_area = image.area()
for it in range(1, 11000):
    image.move()
    area = image.area()
    if area < min_area:
        min_area, min_iter = area, it
    else:
        image.move(-1)
        break

print(f"Part 1:")
image.draw()

res = min_iter
print(f"Part 2: {res}")
assert res == 10888
